haskell.nix

```diff
@@ -0,0 +1,110 @@
+            description = ''Name of the Cabal package'';
+          };
+          root = mkOption {
+            type = types.path;
+            description = ''Path to the Cabal project root'';
+            default = ./.;
+          };
+          overrides = mkOption {
+            type = types.anything;
+            description = ''Overrides for the Cabal project'';
+            default = self: super: { };
+          };
+          modifier = mkOption {
+            type = types.anything;
+            description = ''Modifier for the Cabal project'';
+            default = drv: drv;
+          };
+          baseBuildTools = mkOption {
+            type = types.anything;
+            description = ''Common build tools for Haskell develop; you don't need to set this.'';
+            default = hp: with hp; [
+              cabal-install
+              haskell-language-server
+              ghcid
+              hlint
+              cabal-fmt
+            ];
+          };
+          extraBuildTools = mkOption {
+            type = types.anything;
+            description = ''Extra tools to add to nix-shell'';
+            default = hp: [ ];
+          };
+        };
+      });
+  };
+  config = {
+    perSystem = { config, self', inputs', pkgs, ... }:
+      let
+        cfg = config.haskellProject;
+        inherit (pkgs.lib.lists) optionals;
+
+        # Specify GHC version here. To get the appropriate value, run:
+        #   nix-env -f "<nixpkgs>" -qaP -A haskell.compiler
+        hp = cfg.haskellPackages; # pkgs.haskellPackages; # Eg: pkgs.haskell.packages.ghc921;
+
+        buildTools = cfg.baseBuildTools hp ++ cfg.extraBuildTools hp;
+
+        project =
+          { returnShellEnv ? false
+          , withHoogle ? false
+          }:
+          hp.developPackage {
+            inherit returnShellEnv withHoogle;
+            inherit (cfg) root name overrides;
+            modifier = drv:
+              cfg.modifier (pkgs.haskell.lib.overrideCabal drv (oa: {
+                # All the Cabal-specific overrides go here.
+                # For examples on what is possible, see:
+                #   https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/haskell-modules/lib/compose.nix
+                buildTools = (oa.buildTools or [ ]) ++ optionals returnShellEnv buildTools;
+              }));
+          };
+      in
+      {
+        # Used by `nix build ...`
+        packages = {
+          ${cfg.name} = project { };
+          default = mkDefault self'.packages.${cfg.name};
+        };
+        # Used by `nix run ...`
+        apps = {
+          ${cfg.name} = {
+            type = "app";
+            program = pkgs.lib.getExe self'.packages.default;
+          };
+          default = mkDefault self'.apps.${cfg.name};
+        };
+        # Used by `nix develop ...`
+        devShells = {
+          ${cfg.name} = project { returnShellEnv = true; withHoogle = true; };
+          default = mkDefault self'.devShells.${cfg.name};
```

Comment:
This reminds me that the module does not give us composition for free. This module can only be used for a single package set and a single package.

The module system does have what I like to think of as an imagination, in the form of `extendModules`, which allows it to create a parallel configuration with some changes applied, from which you can use values in your normal configuration. I suppose your user could use this for, say, swapping out `haskellPackages` for a ghcjs-based one, but they'd have to wire up stuff like `packages` themselves.

If you'd like to support multiple package sets, you could wrap all the logic in an `attrsOf (submodule yourModule)`. Sounds simple but requires a bunch of rewiring.

Reply:
@roberth Thanks for your feedback. I've resolved them all except this one, which I don't really understand much. Let me think about it. 

I do, ultimately, want to provide a way to support _multiple_ Haskell packages. Say you have a monorepo of more than 1 Haskell package in sub directories; it should be possible to use this flake module for them. The user would have to manually specify the default I guess; not sure how that can be obviated in a single-package case.

Reply:
> If you'd like to support multiple package sets, you could wrap all the logic in an `attrsOf (submodule yourModule)`. Sounds simple but requires a bunch of rewiring.

Actually, I was thinking of something like this. Maybe you are talking about the same thing. So the new interface I had in mind is to let the user do:

```nix
perSystem = { pkgs, ... }: {
  haskellProjects = {
    "haskell-template" = { 
      buildTools = ...;
    };
    "another-haskell-package" = {};
  };
  defaultPackage = self.another-haskell-package; # etc.
}
```

This will generate for instance `nix build .#another-haskell-package`, but since there is not exactly 1 haskell package, a default will not be generated, so users would have to do it themselves (such as the `defaultPackage` above).

Reply:
c4224f4 implements the above, but it also forces the user to define the `default` outputs, which ideally should have been implicit for single-project use ...

Reply:
I'm mostly satisfied, but there is something funky about this Nix code:

<img width="564" alt="image" src="https://user-images.githubusercontent.com/3998/170889538-a1315883-d70a-4efc-878a-8bc5d1eced53.png">

Reply:
It's a bit repetitive, which is somewhat expected when exposing things from an `attrsOf submodule`.

Btw, the automatic default can be a bit surprising, especially when adding a new project. It's "non-monotonic" behavior. If you add something, you wouldn't expect something else to disappear.

I guess if you could have multiple packages in a single project, fewer people would encounter the problem. That *might* be enough to tip the balance favorably.

Reply:
> Btw, the automatic default can be a bit surprising, especially when adding a new project. It's "non-monotonic" behavior. If you add something, you wouldn't expect something else to disappear.

Perhaps the situation can be improved by always using the first project (`head` of list) as the default value for default attrs. 99% of the use cases I imagine to be using a single package only, so automatic defaults seems valuable - in shaving off 3 lines off flake.nix.

Reply:
Surprises are worse than at least 10 LoC I'd say, but as I've said before, _maybe_ the odds do work out favorably.

You could save 2 user LoC by having an option for specifying which project is the default. What remains then is `defaultHaskellProject = "haskell-template";`

Reply:
> if you could have multiple packages in a single project

I didn't really ask this yet. How do you define a project? I don't think it should be defined equal to a package, because coordinating a single package set from multiple projects is asking for complexity. It would make sense to me to 1:1 with a package set (ie `haskell.packages.x`). Not sure if it would make sense to rename it, as a project might be a somewhat larger concept than a package set.

Reply:
> How do you define a project?

It is whatever [`developPackage`](https://github.com/NixOS/nixpkgs/blob/9bc841fec1c0e8b9772afa29f934d2c7ce57da8e/pkgs/development/haskell-modules/make-package-set.nix#L245) returns, which is either a library or an executable (if there is an `executable` cabal stanza).

I guess I should drop 'project' in favour of 'package' then (to also be consistent with `developPackage` nomenclature). I don't think what's in nixpkgs allows for per-stanza level of find-grained access (which `haskell.nix` seems to provide).

So, what I'm thinking right now:

```nix
{
  haskellFlake = {
    packages = {
      haskell-template = { .. };
      sub-package = { root = ./sub-package; ... };
    };
    defaultPackage = "sub-package";
}
```

Reply:
But then, `package` is too specific, because `haskellFlake.foo` provides not only Nix `packages`, but also `apps` and `devShells`.

Reply:
I haven't used `developPackage` before.
It would make sense to me to use something like `shellFor` instead, which supports working on multiple packages at once. That's what I do in the `hercules-ci-agent` flake (which is still rather a mess actually). If multi-package projects like that aren't your intended audience, I'll have to write a competing module at some point, as part of the cleanup. I'd rather contribute missing bits here, if any.

Reply:
@roberth I don't use multi-package projects personally so I haven't so far thought of this use case. But yea I do think `haskell-flake` should support it, to the point of being useful for repos like `hercules-ci-agent`. Contributions very welcome! 

As for `shellFor`, I haven't used it - but does it actually give you a shell with all sub-packages in it? What happens say if `hercules-ci-api` uses a dependency that is in version conflict with the same in say `hercules-ci-cli`?  

Anyway, let's move this discussion to https://github.com/srid/haskell-flake/issues/7